src/v/kafka/server/offset_translator.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+#include "kafka/types.h"
+#include "model/fundamental.h"
+#include "raft/configuration_manager.h"
+
+namespace kafka {
+/**
+ * Offset translator performs conversion between kafka offsets, which doesn't
+ * include raft configuration batches and log offsets which includes
+ */
+class offset_translator {
+public:
+    explicit offset_translator(const raft::configuration_manager& config_mgr)
+      : _cfg_mgr(config_mgr) {}
+
+    model::offset to_kafka_offset(model::offset o) const {
+        auto d = delta(o);
+        return model::offset(o - d);
+    }
+
+    model::offset from_kafka_offset(
+      model::offset o, model::offset hint = model::offset{}) const {
+        if (o == model::offset::max()) {
+            return o;
+        }
+        model::offset kafka_offset(o);
+        auto current = std::max(hint, kafka_offset);
+        auto next_cfg_it = _cfg_mgr.lower_bound(current);
+        // fast exit, current offset is larger than last config batch
+        // offset, just add delta
+        if (next_cfg_it == _cfg_mgr.end()) {
+            --next_cfg_it;
+            return model::offset(o) + next_cfg_it->second.idx();
+        }
+
+        auto delta = model::offset(next_cfg_it->second.idx() - 1);
+
+        while (current - delta != kafka_offset
+               || current == next_cfg_it->first) {
+            const auto next_cfg_offset = next_cfg_it->first;
+            // largest kafka offset that is achievable before reaching next
+            // control batch
+            const auto max_cfg_ko = next_cfg_offset - delta;
+            if (max_cfg_ko > o) {
+                // offset is between current and next configuration batch,
+                // it is enough to add current delta
+                return kafka_offset + delta;
+            }
+            // skip to the next configuration batch
+            auto prev = next_cfg_it;
+            ++next_cfg_it;
+            delta += model::offset(std::distance(prev, next_cfg_it));
```

Comment:
It looks like `std::distance(prev, next_cfg_it)` is always 1, so it's better to replace it to `delta += model::offset(1)` or to `delta = model::offset(next_cfg_it->second.idx() - 1)`

Reply:
I've refactored this part completely